File: answers/26/main.cpp
#include <vector>
#include <chrono>
#include <iostream>
#include <opencv2/opencv.hpp>
#include <unistd.h>
#include <cmath>
#include "../hist.cpp"

void bilinear_inter(const cv::Mat &src, cv::Mat &dst, float a, float b) {
    // 出力画像
    int dst_h = src.size().height * a;
    int dst_w = src.size().width * b;
    dst = cv::Mat::zeros(cv::Size(dst_h, dst_w), CV_8UC3);

    dst.forEach<cv::Vec3b>([&](cv::Vec3b &pixel, const int pos[]) -> void {
        int _x = pos[1];
        int _y = pos[0];

        int x = floor(_x / b);
        int y = floor(_y / a);

        float dx = _x / b - x;
        float dy = _y / a - y;

        pixel = (1 - dx) * (1 - dy) * src.at<cv::Vec3b>(y, x) +
                dx * (1 - dy) * src.at<cv::Vec3b>(y, x + 1) +
                (1 - dx) * dy * src.at<cv::Vec3b>(y + 1, x) +
                dx * dy * src.at<cv::Vec3b>(y + 1, x + 1);
    });
}

int main() {
    cv::Mat img_orig = cv::imread("questions/dataset/images/imori_256x256.png");
    cv::Mat img_result;
    bilinear_inter(img_orig, img_result, 1.5, 1.5);

    cv::imshow("cpp_origin", img_orig);
    cv::imshow("cpp_result", img_result);
    cv::moveWindow("cpp_origin", 0, -img_orig.size().height);
    cv::moveWindow("cpp_result", img_orig.size().width, -img_orig.size().height);
    cv::waitKey(0);
    return 0;
}

File: answers/26/main.py
import numpy as np
import cv2
import matplotlib.pyplot as plt
import sys
from answers.hist import *


def bilinear_inter(img, a, b):
    h, w, c = img.shape
    out_h = int(h * a)
    out_w = int(w * b)

    xs, ys = np.meshgrid(range(out_w), range(out_h)) # output image index

    _xs = np.floor(xs / b).astype(int) # original x
    _ys = np.floor(ys / a).astype(int) # original y

    dx = xs / b - _xs
    dy = ys / a - _ys

    dx = np.repeat(np.expand_dims(dx, axis=-1), c, axis=-1) # repeat channel
    dy = np.repeat(np.expand_dims(dy, axis=-1), c, axis=-1) # repeat channel

    _xs1p = np.minimum(_xs + 1, w - 1)
    _ys1p = np.minimum(_ys + 1, h - 1)


    out = (1 - dx) * (1 - dy) * img[_ys, _xs] + dx * (1 - dy) * img[_ys, _xs1p] + \
                    (1 - dx) * dy * img[_ys1p, _xs] + dx * dy * img[_ys1p, _xs1p]

    return np.clip(out, 0, 255).astype(np.uint8)

if __name__ == '__main__':
    img_orig = cv2.imread('questions/dataset/images/imori_256x256.png')
    img_result = bilinear_inter(img_orig, a=1.5, b=1.5)

    cv2.imshow('python_origin', img_orig)
    cv2.imshow('python_result', img_result)
    cv2.moveWindow('python_result', img_orig.shape[1], 0)
    cv2.waitKey(0)
